File: packages/websearch/src/engines/rank.ts
import type { EngineClass, WebSearchResultItem } from "../types.js";

/**
 * Reciprocal Rank Fusion (RRF) with engine weights — the merge ranker used
 * when the fallback chain gathers results from more than one engine.
 *
 * Why RRF: the engines' native signals are NOT comparable (Marginalia's
 * `quality` float, Tavily's 0–1 `score`, Mojeek's bare rank). Fusing on RANK
 * sidesteps that, and is the established metasearch approach (SearXNG,
 * Elasticsearch hybrid search). For each result URL:
 *
 *     fused(d) = Σ over engines e that returned d:  weight(e) / (K + rank_e(d))
 *
 * - `rank_e(d)` is 0-based position in engine e's list.
 * - K is small (10) because our lists are short (≤20), unlike the TREC default
 *   of 60 tuned for thousand-item lists.
 * - Two engines returning the same URL SUM their contributions → a consensus
 *   boost ("two independent indexes agree" is the strongest cheap relevance
 *   signal). The A/B (scenario 3) showed this is the main reorder win.
 * - Engine weights (general > niche > vertical; keyed providers highest) keep
 *   the encyclopedic/indie backstop from outranking broad web purely because
 *   the leader was short (A/B scenario 4).
 *
 * Determinism: ties break by (1) higher single best per-engine weight/rank,
 * then (2) original insertion order, so the output is stable for a given set
 * of engine lists.
 */

export const RRF_K = 10;

/** Default per-class engine weights. Keyed providers rank above keyless. */
export const ENGINE_WEIGHTS: Readonly<Record<EngineClass, number>> = {
  general: 1.0,
  niche: 0.8,
  vertical: 0.6,
};

/** Brave/Tavily (keyed, official) get a small premium over keyless general. */
export const KEYED_ENGINE_WEIGHT = 1.2;
const KEYED_ENGINES = new Set(["brave", "tavily"]);

export function engineWeight(name: string, engineClass: EngineClass): number {
  if (KEYED_ENGINES.has(name)) return KEYED_ENGINE_WEIGHT;
  return ENGINE_WEIGHTS[engineClass];
}

/** One engine's contribution to a URL: which engine, at what 0-based rank. */
export interface RankOccurrence {
  readonly engine: string;
  readonly engineClass: EngineClass;
  readonly rank: number;
}

/** A candidate URL accumulated across engines, before fusion. */
export interface FusionCandidate {
  /** The result item to emit (first engine to surface the URL owns the fields). */
  item: WebSearchResultItem;
  /** Every (engine, rank) that returned this URL, in insertion order. */
  occurrences: RankOccurrence[];
  /** Insertion order index, for a stable final tiebreak. */
  readonly order: number;
}

export interface FusedResult {
  readonly item: WebSearchResultItem;
  readonly score: number;
  /** Contributing engine names in best-rank-first order (for `source`). */
  readonly sources: readonly string[];
}

/** Compute the fused RRF score for one candidate. */
export function fusedScore(occ: readonly RankOccurrence[]): number {
  let s = 0;
  for (const o of occ) s += engineWeight(o.engine, o.engineClass) / (RRF_K + o.rank);
  return s;
}

/**
 * Fuse accumulated candidates into a ranked list (best-first). Pure function:
 * given the same candidates it returns the same order.
 */
export function fuseRrf(
  candidates: readonly FusionCandidate[],
): FusedResult[] {
  const scored = candidates.map((c) => {
    // sources ordered by the engine's own rank (best rank first)
    const sources = [...c.occurrences]
      .sort((a, b) => a.rank - b.rank)
      .map((o) => o.engine);
    return { item: c.item, score: fusedScore(c.occurrences), sources, order: c.order };
  });
  scored.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    // Tie: more engines agreeing wins, then original insertion order.
    if (b.sources.length !== a.sources.length) {
      return b.sources.length - a.sources.length;
    }
    return a.order - b.order;
  });
  return scored.map(({ item, score, sources }) => ({ item, score, sources }));
}

File: crates/websearch/src/engines/rank.rs
//! Reciprocal Rank Fusion (RRF) + engine weights — the merge ranker used when
//! the fallback chain gathers results from more than one engine. Mirrors the
//! TS `engines/rank.ts`.
//!
//! `fused(d) = sum over engines e returning d of weight(e) / (K + rank_e(d))`
//!
//! Fusing on RANK sidesteps that the engines' native scores aren't comparable
//! (Marginalia `quality` vs Tavily 0–1 `score` vs Mojeek bare rank). K is small
//! (10) for our short lists. Two engines on the same URL SUM contributions (a
//! consensus boost). Engine weights (general > niche > vertical; keyed highest)
//! keep a backstop from outranking broad web. Deterministic ties: score, then
//! number of agreeing engines, then insertion order.

use crate::engine::EngineClass;
use crate::types::WebSearchResultItem;

pub const RRF_K: f64 = 10.0;

/// Brave/Tavily (keyed, official) premium over keyless general.
pub const KEYED_ENGINE_WEIGHT: f64 = 1.2;

pub fn engine_class_weight(class: EngineClass) -> f64 {
    match class {
        EngineClass::General => 1.0,
        EngineClass::Niche => 0.8,
        EngineClass::Vertical => 0.6,
    }
}

pub fn engine_weight(name: &str, class: EngineClass) -> f64 {
    if name == "brave" || name == "tavily" {
        KEYED_ENGINE_WEIGHT
    } else {
        engine_class_weight(class)
    }
}

/// One engine's contribution to a URL: which engine, at what 0-based rank.
#[derive(Clone)]
pub struct RankOccurrence {
    pub engine: String,
    pub class: EngineClass,
    pub rank: usize,
}

/// A candidate URL accumulated across engines, before fusion.
pub struct FusionCandidate {
    /// The item to emit (first engine to surface the URL owns the fields).
    pub item: WebSearchResultItem,
    /// Every (engine, rank) that returned this URL, in insertion order.
    pub occurrences: Vec<RankOccurrence>,
    /// Insertion order index, for a stable final tiebreak.
    pub order: usize,
}

pub struct FusedResult {
    pub item: WebSearchResultItem,
    pub score: f64,
    /// Contributing engine names in best-rank-first order (for `source`).
    pub sources: Vec<String>,
}

pub fn fused_score(occ: &[RankOccurrence]) -> f64 {
    occ.iter()
        .map(|o| engine_weight(&o.engine, o.class) / (RRF_K + o.rank as f64))
        .sum()
}

/// Fuse accumulated candidates into a ranked list (best-first). Pure: same
/// candidates → same order.
pub fn fuse_rrf(candidates: Vec<FusionCandidate>) -> Vec<FusedResult> {
    let mut scored: Vec<(FusedResult, usize)> = candidates
        .into_iter()
        .map(|c| {
            let score = fused_score(&c.occurrences);
            let mut occ = c.occurrences;
            occ.sort_by_key(|o| o.rank);
            let sources = occ.into_iter().map(|o| o.engine).collect();
            (
                FusedResult {
                    item: c.item,
                    score,
                    sources,
                },
                c.order,
            )
        })
        .collect();

    scored.sort_by(|a, b| {
        // Higher score first.
        b.0.score
            .partial_cmp(&a.0.score)
            .unwrap_or(std::cmp::Ordering::Equal)
            // Tie: more agreeing engines first.
            .then(b.0.sources.len().cmp(&a.0.sources.len()))
            // Then original insertion order.
            .then(a.1.cmp(&b.1))
    });

    scored.into_iter().map(|(r, _)| r).collect()
}
